### PR TITLE
pull request for Mode functions

### LINKED ISSUE #73: Creating a new function Mode
Description: 
The mode function determines the most frequently occurring value in a list of numbers. If there is no unique mode (e.g., in a tie), it returns an appropriate error message

### PR DESCRIPTION
**Pull request by Oghenekevwe123**
---
name: solution review 
about: A template PR for code review with a checklist
---

# Behavior

## Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [ ] All tests pass
- [ ] There are tests for defensive assertions
- [ ] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
  - _It's ok to have extra helper functions if necessary, like with mergesort_
- [x] The function has correct type annotations
- [x] The function is not called at the top level of the function file 
  - _Recursive solutions **can** call the function from **inside** the function body_

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in any docstrings or tests
- [x] Comments explain the _strategy_, **not** the _implementation_
- [ ] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [ ] The code passes the formatting checks
- [ ] The code passes all Ruff linting checks
- [ ] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [ ] The implementation is as simple as possible given the strategy
- [ ] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

**Comment by Bigbadwolf21** [comment]
I think this looks good to me.
Kindly reformat the code to pass all the CI Checks.

**Comment by Bigbadwolf21** [comment]
Well done Kevwe

File: solutions/tests/test_mode_of_numbers.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        expected = [-2]
+        self.assertEqual(actual, expected)
+
+    def test_Mode_7(self):
+        """It should return an empty list for a list where no number repeats equally"""
+        actual = mode_of_numbers([1, 2, 1, 2])
+        expected = []
+        self.assertEqual(actual, expected)
+
+    def test_Mode_8(self):
```
Comment (Yevheniia-Rudenko) [inline solutions/tests/test_mode_of_numbers.py:65]: The test test_Mode_8 is failing because the function mode_of_numbers returns [1, 2, 3, 4, 5, 6] instead of the expected empty list []. This happens because the function does not check if the maximum frequency is greater than 1.

Suggestion:
Add a condition to check if the maximum frequency (max_frequency) is equal to 1. If all elements in the list are unique, the function should return an empty list [].

**Comment by Yevheniia-Rudenko** [comment]
Hey @Oghenekevwe123 ,

I noticed a couple of things that might need fixing:

**Ruff CI Issues:**
The CI checks are failing due to some style issues flagged by Ruff. You can run ruff check . locally to see what needs fixing and resolve them.
**Failing Test (test_Mode_8):**
The test test_Mode_8 isn’t passing because the function doesn’t handle the case where all numbers in the list are unique. If there’s no mode (i.e., all numbers appear only once), the function should return an empty list []. Adding a simple check like this should fix it:
if max_frequency == 1:
    return []
**Test Names:**
It might be helpful to rename the tests to make their purpose clearer. For example:
test_Mode_8 → test_returns_empty_list_for_unique_numbers
test_Mode_9 → test_returns_multiple_modes_when_tied
**Lastly:**
I know this feedback is coming after the deadline, but I hope it’s still useful for improving the project in the future! Let me know if I can help with anything else. 😊
